### PR TITLE
Feature: Indents

### LINKED ISSUE #140: Feature request: Extend Vaadin RTE to add indentations
Support Ticket #3431

Enable indent functionality and add buttons to toolbar:

Screenshot 2020-07-13 at 14 04 36

This will require new Lumo icons?

Will need to consider how this works with RTL.

### PR DESCRIPTION
**Pull request by niiyeboah**
Closes #140

### CODE REVIEW

**Comment by niiyeboah** [comment]
The temp icons are on my [Figma](https://www.figma.com/file/fmGrfdOLyOwdkQdybZ0Vly/Indent-Icons?node-id=0%3A1).

**Review by yuriy-fix — CHANGES_REQUESTED** [review]
Do we have some specifications for the expected keyboard hotkeys for indentations? 🤔 
For me it's a bit unexpected that <kbd>Tab</kbd> applies indent styles in `lists`, but in i.e. paragraphs it's adding `\t` 😅 
It's also not obvious that decreasing of the indentation should be done with <kbd>Backspace</kbd>. Personally I have moved the focus to the toolbar ~10 times with <kbd>Shift</kbd> + <kbd>Tab</kbd> while testing instead of changing the indentation.

**Comment by rolfsmeds** [comment]
> Do we have some specifications for the expected keyboard hotkeys for indentations? 🤔
> For me it's a bit unexpected that Tab applies indent styles in `lists`, but in i.e. paragraphs it's adding `\t` 😅
> It's also not obvious that decreasing of the indentation should be done with Backspace. Personally I have moved the focus to the toolbar ~10 times with Shift + Tab while testing instead of changing the indentation.

This has not been explicitly discussed, and as such, my expectation is that it should work as in standard Quill, which seems to be just like most rich text editors out there:
- Tab key adds \t in paragraphs and indents lists.
- BOTH Shift-Tab AND backspace remove \t and unindent lists.

If there are platform-specific reasons to deviate from these, we need to discuss.

File: src/vaadin-rich-text-editor-content-styles.html [hunk 1]
```diff
@@ -66,6 +66,189 @@
...
+
+      :host([dir="rtl"]) .ql-editor li.ql-indent-8:not(.ql-align-left) {
+        margin-right: 25.5rem;
+      }
+
+      .ql-editor li.ql-indent-9 {
+        list-style-type: disc;
+      }
+
+      :host(:not([dir="rtl"])) .ql-editor .ql-indent-9 {
```
[the remark this answers is not in the document]
Comment (rolfsmeds) [reply src/vaadin-rich-text-editor-content-styles.html:236]: No limitation has been specified. Is Quill somehow limited to 8?

File: src/vaadin-rich-text-editor.html [hunk 2]
```diff
@@ -847,19 +876,30 @@
           const editor = this.shadowRoot.querySelector(`.${className}`);
           let content = editor.innerHTML;
 
+          // Convert quill indent styles to nested lists
+          // https://github.com/quilljs/quill/issues/979
+          content = this._convertLists(content);
```
[the remark this answers is not in the document]
Comment (rolfsmeds) [reply src/vaadin-rich-text-editor.html:881]: Yes, nested lists are expected to work correctly.

File: src/vaadin-rich-text-editor-toolbar-styles.html [hunk 3]
```diff
@@ -37,16 +37,8 @@
         transform: translate(-50%, -50%);
       }
 
-      [part~="toolbar-button-undo"]::before,
```
[the remark this answers is not in the document]
Comment (niiyeboah) [reply src/vaadin-rich-text-editor-toolbar-styles.html:40]: Do these styles get used? When I was testing it seemed that the icons were coming from lumo-styles or material-styles

[the remark this answers is not in the document]
Comment (niiyeboah) [reply src/vaadin-rich-text-editor-toolbar-styles.html:40]: I can put them back, but it seemed like they were unnecessary

File: src/vaadin-rich-text-editor-content-styles.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuriy-fix) [reply src/vaadin-rich-text-editor-content-styles.html:236]: As discussed with @rolfsmeds, I guess for this particular case we can go with the limitation to `ql-indent-8` as Quill does it.
P.S. So `ql-indent-9` can be removed.

**Comment by yuriy-fix** [comment]
> This has not been explicitly discussed, and as such, my expectation is that it should work as in standard Quill, which seems to be just like most rich text editors out there:
> 
> * Tab key adds \t in paragraphs and indents lists.
> * BOTH Shift-Tab AND backspace remove \t and unindent lists.
> 
> If there are platform-specific reasons to deviate from these, we need to discuss.

As discussed with @rolfsmeds, for this case we can go with behavioral altering change, so that <kbd>Shift</kbd> + <kbd>Tab</kbd> won't anymore move focus back to the toolbar and <kbd>Esc</kbd> will do that instead. So <kbd>Shift</kbd> + <kbd>Tab</kbd> will now remove \t and unindent lists (as well as <kbd>Backspace</kbd>)

File: src/vaadin-rich-text-editor.html [hunk 4]
```diff
@@ -1002,6 +1042,66 @@
...
+        /** @private */
+        _convertLists(content) {
+          const textNode = (str) => document.createTextNode(str);
+          const tempEl = document.createElement('div');
+          tempEl.setAttribute('style', 'display: none;');
+          tempEl.innerHTML = content;
+
+          ['ul', 'ol'].forEach((type) => {
+            const startTag = `-start-${type}-`;
+            const endTag = `-end-${type}-`;
```
Comment (web-padawan) [inline src/vaadin-rich-text-editor.html:1054]: This only allows nesting lists of the same, i.e. not possible to nest `<ul>` lists inside `<ol>` and vice versa.

The problem is how we detect which type of list does the `<li>` element belong to. 
Quill master branch uses `li[data-list=ordered]` and `li[data-list=bullet]`, maybe we could do the same.

Then we would need to take that into account when converting lists.

Comment (web-padawan) [reply src/vaadin-rich-text-editor.html:1054]: Another thing that I noticed at Quill master branch is that they [always use `<ol>`](https://github.com/quilljs/quill/blob/e9f628b5562e78126291bb25f63d0dfc3f6c6f86/formats/list.js#L7) tag.

Then the markup for the bullet lists looks like this:

```html
<ol>
  <li data-list="ordered">0123</li>
  <li data-list="ordered">5678</li>
</ol>
```

I'm not sure if upgrading Quill to 2.0 would make this feature easier to implement or not.

**Comment by yuriy-fix** [comment]
Closing as stale.

**Pull request closed without merging**